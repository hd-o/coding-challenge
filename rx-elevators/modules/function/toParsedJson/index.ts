export type ToParsedJson <ParsedType = unknown> = <P = ParsedType> (v: unknown) => P

export const useToParsedJson = (): ToParsedJson => (v) =>
  JSON.parse(JSON.stringify(v, null, 2))
