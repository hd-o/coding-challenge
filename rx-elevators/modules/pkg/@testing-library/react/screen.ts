import { screen } from '@testing-library/react'

export const useTestingLibraryReactScreen = (): typeof screen => screen
