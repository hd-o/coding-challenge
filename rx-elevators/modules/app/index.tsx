import { FC, useContext, useEffect, useMemo } from 'react'
import { BuildingCtx } from '../building'
import { useStartup } from '../startup'
import { useResolve } from '../util/useResolve'

export const App: FC<{}> = () => {
  const Building = useContext(BuildingCtx)
  const startup = useResolve(useStartup)

  /**
   * Startup subscription needs to happen before rendering
   * child components, so subscribers to shared observables
   * receive the first emitted value
   */
  const subscription = useMemo(() => startup(), [startup])

  useEffect(() => () => subscription.unsubscribe())

  return (
    <div style={{ paddingTop: 20, width: 800, margin: '0 auto' }}>
      <div className='ui container'>
        <Building />
      </div>
    </div>
  )
}
