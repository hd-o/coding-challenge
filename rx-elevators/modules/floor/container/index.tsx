import { createContext, FC, useContext } from 'react'
import { useElevatorId$ } from '../../elevator/id/stream'
import { FloorPanelCallerCtx } from '../../floor/panel/caller'
import { useRamdaReverse } from '../../pkg/ramda/reverse'
import { useRamdaTimes } from '../../pkg/ramda/times'
import { useResolve } from '../../util/useResolve'
import { useStreamFn } from '../../util/useStreamFn'
import { floorHeight } from '../height'
import { FloorNumber } from '../number'
import { useFloorNumber$ } from '../number/stream'
import { useFloorRequestEvent$ } from '../request/event/stream'

const FloorContainer: FC<{}> = () => {
  const FloorPanelCaller = useContext(FloorPanelCallerCtx)

  const floorRequestEvent$ = useResolve(useFloorRequestEvent$)
  const reverse = useResolve(useRamdaReverse)
  const times = useResolve(useRamdaTimes)

  const elevatorIds = useStreamFn(useElevatorId$)
  const floorNumbers = useStreamFn(useFloorNumber$)

  const floorCells = times(i => <td key={i} />, elevatorIds.length)

  const handleFloorCall = (number: FloorNumber) =>
    () => floorRequestEvent$.next(number)

  return <>
    {
      reverse(floorNumbers).map(number => (
        <tr key={number} style={{ height: floorHeight }}>
          <td className='two wide' style={{ textAlign: 'center' }}>
            <FloorPanelCaller
              floor={number}
              onClick={handleFloorCall(number)}
              testidPrefix='floor-panel'
            />
          </td>
          {floorCells}
        </tr>
      ))
    }
  </>
}

export const FloorContainerCtx = createContext(FloorContainer)
