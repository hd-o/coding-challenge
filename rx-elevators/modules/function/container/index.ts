import { createContext } from 'react'

type FunctionContainerValueConstructor <V = any> = (c: FunctionContainer) => V

export type FnCtor <V = any> = FunctionContainerValueConstructor<V>

class FunctionContainer {
  constructor (private readonly _parent?: FnC) {}

  private _map = new WeakMap()

  private _resolve <F extends FnCtor> (fn: F): ReturnType<typeof fn> | undefined {
    if (this._map.has(fn)) return this._map.get(fn)
    if (this._parent === undefined) return undefined
    return this._parent._resolve(fn)
  }

  readonly childContainer = (): FnC => new FunctionContainer(this)

  readonly register = <F extends FnCtor> (fn: F, value: FnCtor<ReturnType<typeof fn>>): void => {
    this._map.set(fn, value(this))
  }

  readonly reset = (): void => {
    this._map = new WeakMap()
  }

  readonly resolve = <F extends FnCtor> (fn: F): ReturnType<typeof fn> => {
    const resolved = this._resolve(fn)
    if (resolved !== undefined) return resolved
    const value = fn(this)
    this._map.set(fn, value)
    return value
  }
}

export type FnC = FunctionContainer

export const fnContainer = new FunctionContainer()
export const FnContainerCtx = createContext(fnContainer)
