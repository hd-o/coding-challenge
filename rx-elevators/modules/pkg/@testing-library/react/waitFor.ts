import { waitFor } from '@testing-library/react'

export const useTestingLibraryReactWaitFor = (): typeof waitFor => waitFor
