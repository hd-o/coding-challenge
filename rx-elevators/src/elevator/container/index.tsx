import { useStreamFn } from '/src/util/useStreamFn'
import { createContext, FC, useContext } from 'react'
import { ElevatorCtx } from '../'
import { useElevatorDoorState$Map$ } from '../door/state/stream/map'
import { useElevatorId$ } from '../id/stream'
import { useElevatorPosition$Map$ } from '../position/stream/map/stream'

const ElevatorContainer: FC<{}> = () => {
  const Elevator = useContext(ElevatorCtx)

  const elevatorIds = useStreamFn(useElevatorId$)
  const doorState$Map = useStreamFn(useElevatorDoorState$Map$)
  const position$Map = useStreamFn(useElevatorPosition$Map$)

  return <>
    {
      elevatorIds.map((id) => {
        const doorState$ = doorState$Map.get(id)
        const position$ = position$Map.get(id)
        if (position$ === undefined) return null
        if (doorState$ === undefined) return null
        return <Elevator
          doorState$={doorState$}
          id={id}
          key={String(id)}
          position$={position$}
        />
      })
    }
  </>
}

export const ElevatorContainerCtx = createContext(ElevatorContainer)
