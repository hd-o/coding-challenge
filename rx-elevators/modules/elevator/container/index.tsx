import { createContext, FC, useContext } from 'react'
import { ElevatorCtx } from '../'
import { floorHeight } from '../../floor/height'
import { FloorNumber } from '../../floor/number'
import { useFloorNumber$ } from '../../floor/number/stream'
import { FloorPanelCallerCtx } from '../../floor/panel/caller'
import { useFloorRequestEvent$ } from '../../floor/request/event/stream'
import { useRamdaReverse } from '../../pkg/ramda/reverse'
import { useRamdaTimes } from '../../pkg/ramda/times'
import { SemanticUiTableCtx } from '../../pkg/semantic-ui/Table'
import { useResolve } from '../../util/useResolve'
import { useStreamFn } from '../../util/useStreamFn'
import { useElevatorDoorState$Map$ } from '../door/state/stream/map'
import { useElevatorId$ } from '../id/stream'
import { useElevatorPosition$Map$ } from '../position/stream/map/stream'

const ElevatorContainer: FC<{}> = () => {
  const Elevator = useContext(ElevatorCtx)
  const FloorPanelCaller = useContext(FloorPanelCallerCtx)
  const Table = useContext(SemanticUiTableCtx)

  const floorRequestEvent$ = useResolve(useFloorRequestEvent$)
  const reverse = useResolve(useRamdaReverse)
  const times = useResolve(useRamdaTimes)

  const elevatorIds = useStreamFn(useElevatorId$)
  const floorNumbers = useStreamFn(useFloorNumber$)

  const doorState$Map = useStreamFn(useElevatorDoorState$Map$)
  const position$Map = useStreamFn(useElevatorPosition$Map$)

  const elevators = elevatorIds.map((id) => {
    const doorState$ = doorState$Map.get(id)
    const position$ = position$Map.get(id)
    if (position$ === undefined) return null
    if (doorState$ === undefined) return null
    return <Elevator
      doorState$={doorState$}
      id={id}
      key={String(id)}
      position$={position$} />
  })

  const newTableCell = (i: number): JSX.Element => <Table.Cell key={i} />

  const elevatorFloorCells = times(newTableCell, elevatorIds.length)

  const handleFloorCall = (number: FloorNumber) =>
    () => floorRequestEvent$.next(number)

  const floors = reverse(floorNumbers).map(number => (
    <Table.Row key={number} style={{ height: floorHeight }}>
      <Table.Cell width='2' textAlign='center'>
        <FloorPanelCaller
          floor={number}
          onClick={handleFloorCall(number)}
        />
      </Table.Cell>
      {elevatorFloorCells}
    </Table.Row>
  ))

  return (
    <Table fixed compact celled definition unstackable>
      <Table.Body>
        {floors}
      </Table.Body>
      <Table.Footer fullWidth style={{ height: floorHeight }}>
        <Table.Row>
          <Table.HeaderCell />
          {elevators}
        </Table.Row>
      </Table.Footer>
    </Table>
  )
}

export const ElevatorContainerCtx = createContext(ElevatorContainer)
