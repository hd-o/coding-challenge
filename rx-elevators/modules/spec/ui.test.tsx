/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom'
import { fireEvent, render, screen } from '@testing-library/react'
import { App } from '../app'
import { fnContainer } from '../function/container'
import { useFloorWasRequested } from './util/ui/floorWasRequested'
import { useGetFloorCaller } from './util/ui/getFloorCaller'

let floorWasRequested: ReturnType<typeof useFloorWasRequested>
let getFloorCaller: ReturnType<typeof useGetFloorCaller>

beforeEach(() => {
  fnContainer.reset()
  floorWasRequested = fnContainer.resolve(useFloorWasRequested)
  getFloorCaller = fnContainer.resolve(useGetFloorCaller)
})

describe('UI render', () => {
  test('App snapshot', async () => {
    const app = render(<App />)
    expect(app.asFragment()).toMatchSnapshot('app-render')
  })
  test('floor panel request', async () => {
    render(<App />)
    const caller = getFloorCaller('floor-panel', 1)
    fireEvent.click(caller)
    expect(await floorWasRequested(1)).toBe(true)
  })
  test('elevator panel request', async () => {
    render(<App />)
    const [elevatorPanel] = screen.getAllByTestId(/elevator-\d-panel/)
    const { testid = '' } = elevatorPanel.dataset
    const caller = getFloorCaller(testid, 1)
    fireEvent.click(caller)
    expect(await floorWasRequested(1, testid)).toBe(true)
  })
})
