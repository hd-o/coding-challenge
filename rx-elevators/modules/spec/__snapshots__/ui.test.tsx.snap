// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI render App snapshot: app-render 1`] = `
<DocumentFragment>
  <div
    style="padding-top: 20px; width: 800px; margin: 0px auto;"
  >
    <div
      class="ui container"
    >
      <table
        class="ui compact celled definition unstackable table"
      >
        <tbody>
          <tr
            style="height: 100px;"
          >
            <td
              class="two wide"
              style="text-align: center;"
            >
              <button
                class="ui button info"
                data-testid="floor-panel__floor-caller-4"
                style="padding: 10px; margin-top: 5px;"
              >
                4
              </button>
            </td>
            <td />
            <td />
            <td />
          </tr>
          <tr
            style="height: 100px;"
          >
            <td
              class="two wide"
              style="text-align: center;"
            >
              <button
                class="ui button info"
                data-testid="floor-panel__floor-caller-3"
                style="padding: 10px; margin-top: 5px;"
              >
                3
              </button>
            </td>
            <td />
            <td />
            <td />
          </tr>
          <tr
            style="height: 100px;"
          >
            <td
              class="two wide"
              style="text-align: center;"
            >
              <button
                class="ui button info"
                data-testid="floor-panel__floor-caller-2"
                style="padding: 10px; margin-top: 5px;"
              >
                2
              </button>
            </td>
            <td />
            <td />
            <td />
          </tr>
          <tr
            style="height: 100px;"
          >
            <td
              class="two wide"
              style="text-align: center;"
            >
              <button
                class="ui button info"
                data-testid="floor-panel__floor-caller-1"
                style="padding: 10px; margin-top: 5px;"
              >
                1
              </button>
            </td>
            <td />
            <td />
            <td />
          </tr>
          <tr
            style="height: 100px;"
          >
            <td
              class="two wide"
              style="text-align: center;"
            >
              <button
                class="ui button info"
                data-testid="floor-panel__floor-caller-0"
                style="padding: 10px; margin-top: 5px;"
              >
                0
              </button>
            </td>
            <td />
            <td />
            <td />
          </tr>
        </tbody>
        <tfoot
          class="full-width"
        >
          <tr
            style="height: 100px;"
          >
            <th />
            <th
              style="text-align: center; position: relative; overflow: visible;"
            >
              <div
                data-testid="elevator-1-panel"
              >
                <button
                  class="ui button info"
                  data-testid="elevator-1-panel__floor-caller-0"
                  style="padding: 10px; margin-top: 5px;"
                >
                  0
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-1-panel__floor-caller-1"
                  style="padding: 10px; margin-top: 5px;"
                >
                  1
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-1-panel__floor-caller-2"
                  style="padding: 10px; margin-top: 5px;"
                >
                  2
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-1-panel__floor-caller-3"
                  style="padding: 10px; margin-top: 5px;"
                >
                  3
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-1-panel__floor-caller-4"
                  style="padding: 10px; margin-top: 5px;"
                >
                  4
                </button>
              </div>
              <div
                style="height: 100px; width: 100%; background: rgb(187, 187, 187); position: absolute; bottom: 100px; left: 0px;"
              >
                <div
                  style="display: flex; height: 100%; width: 100%; justify-content: space-between;"
                >
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                </div>
              </div>
            </th>
            <th
              style="text-align: center; position: relative; overflow: visible;"
            >
              <div
                data-testid="elevator-2-panel"
              >
                <button
                  class="ui button info"
                  data-testid="elevator-2-panel__floor-caller-0"
                  style="padding: 10px; margin-top: 5px;"
                >
                  0
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-2-panel__floor-caller-1"
                  style="padding: 10px; margin-top: 5px;"
                >
                  1
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-2-panel__floor-caller-2"
                  style="padding: 10px; margin-top: 5px;"
                >
                  2
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-2-panel__floor-caller-3"
                  style="padding: 10px; margin-top: 5px;"
                >
                  3
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-2-panel__floor-caller-4"
                  style="padding: 10px; margin-top: 5px;"
                >
                  4
                </button>
              </div>
              <div
                style="height: 100px; width: 100%; background: rgb(187, 187, 187); position: absolute; bottom: 100px; left: 0px;"
              >
                <div
                  style="display: flex; height: 100%; width: 100%; justify-content: space-between;"
                >
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                </div>
              </div>
            </th>
            <th
              style="text-align: center; position: relative; overflow: visible;"
            >
              <div
                data-testid="elevator-3-panel"
              >
                <button
                  class="ui button info"
                  data-testid="elevator-3-panel__floor-caller-0"
                  style="padding: 10px; margin-top: 5px;"
                >
                  0
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-3-panel__floor-caller-1"
                  style="padding: 10px; margin-top: 5px;"
                >
                  1
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-3-panel__floor-caller-2"
                  style="padding: 10px; margin-top: 5px;"
                >
                  2
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-3-panel__floor-caller-3"
                  style="padding: 10px; margin-top: 5px;"
                >
                  3
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-3-panel__floor-caller-4"
                  style="padding: 10px; margin-top: 5px;"
                >
                  4
                </button>
              </div>
              <div
                style="height: 100px; width: 100%; background: rgb(187, 187, 187); position: absolute; bottom: 100px; left: 0px;"
              >
                <div
                  style="display: flex; height: 100%; width: 100%; justify-content: space-between;"
                >
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                </div>
              </div>
            </th>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
